File: ruby/mixins.rb
module Attributable
  attr_accessor :attributes

  def setAttribute(key, value:)
    attributes[key] = value
  end

  def formattedAttributes
    ret = ""

    attributes.each do |key, value|
      ret += "| key: #{key}\n  value: #{value} |\n"
    end

    ret
  end
end

class BaseModel
  include Attributable

  def initialize
    self.attributes = {}
  end
end

class SomeModel < BaseModel
  def printFormattedAttributes
    puts formattedAttributes()
  end
end

obj = SomeModel.new()
obj.setAttribute("car", value: "Mazda")
obj.setAttribute("year", value: 2011)

obj.printFormattedAttributes()

File: js/mixins.js
"use strict";

let Attributable = (superclass) => class extends superclass {
  setAttribute(key, value) {
    this.attributes[key] = value;
  }

  formattedAttributes() {
    let ret = "";

    for (var key in this.attributes) {
      ret += `| key: ${key}\n  value: ${this.attributes[key]} |\n`;
    }

    return ret;
  }
};

class BaseModel extends Attributable(class {}) {
  constructor() {
    super();
    this.attributes = {};
  }
}

class SomeModel extends BaseModel {
  printFormattedAttributes() {
    console.log(this.formattedAttributes());
  }
}

let obj = new SomeModel();
obj.setAttribute("car", "Mazda");
obj.setAttribute("year", 2011);

obj.printFormattedAttributes();
